### ISSUE TITLE
Move to web3.py

### ISSUE THREAD

**Issue by palango** [raiden-network/raiden#1293]
`pyethereum` isn't maintained properly any more and we should move away from it. I created a proof of concept on converting Raiden to use the web3 libraries. So far I got rid of most calls to our own RPC client (see [here](https://github.com/palango/raiden/commit/e49af6e6861856f93a47217c363e144b555513cd), and the [build](https://travis-ci.org/palango/raiden/builds/348175792)), but before I invest more time I'd like to know if you agree to the plan below.

I see the following two advantages in using web3 over the current situation:
- It's maintained and maintainers are responsible (I got some necessary patches merged in less than a day)
- We get a lot of goodies for free
  - [Caching](https://web3py.readthedocs.io/en/latest/middleware.html#cache)
  - [gas price strategies](https://web3py.readthedocs.io/en/latest/gas_price.html)
  - [Check if chain is synched](https://web3py.readthedocs.io/en/latest/middleware.html#stalecheck)
  - [Reconnection](https://web3py.readthedocs.io/en/latest/middleware.html#httprequestretry) (not in the form we have currently, but the basics are there)

The next steps would be:
1) Remove the `ContractProxy` and use `web3.contract.Contract` instead
2) Rework filtering to use the web3 APIs
3) Remove our own implementations of caching, reconnection and gas price strategies and replace them with the web3 implementations.
4) Rework testing to use eth-tester

### Required
- [x] Replace `slogging` with `structlog`: ~~#1417~~ 
- [x] Remove usages of the `ethereum`  module: ~~#1418~~, ~~#1421~~
- [x] Replace internal RPC calls with calls using web3: #1427 
- [x] Remove `smart_contracts` and `property` tests: ~~#1437~~
- [x] Replace `smoketest` genesis creation: ~~#1446~~
- [x] Replace or remove blockchain cache: ~~#1449~~
- [x] Remove usage of the `ContractTranslator`: ~~#1450~~ 
- [x] Remove calls to `_solidity` module: ~~#1463~~
- [x] Remove last usages of `pyethereum`: #1427 

### Later refactorings
- [ ] Rework filtering to use the web3 APIs: #1466
- [ ] Remove our own implementations of caching and gas price strategies and replace them with the web3 implementations: #1467

**Comment by LefterisJP** [comment]
web3.py is definitely the way to go but what I am afraid of is the potential downtime concerning the integration. Do you have a ballpark figure on how many man-days would it take to do so?

**Comment by palango** [comment]
A rough estimate is 5 days for 1, 2 and 3. I guess 4) can be done parallel with other changes and 5) isn't really urgent.

**Comment by andrevmatos** [comment]
I think this issue is duplicate of #1019. I'm looking at doing it together with the transport (as it may be of good use for identity/signature checks in the Matrix client).

**Comment by palango** [comment]
Yeah, this is a duplicate.

What usage do you see for identity/signature checks for matrix? I don't think it's necessary to couple these two problems together.

**Comment by LefterisJP** [comment]
@andrevmatos We don't want to use any matrix specific features. As we don't want to be bound to a particular transport layer. Please keep that in mind.

**Comment by andrevmatos** [comment]
No, not matrix for identity.. web3.py for identity on matrix. I use it to interact with eth node, sign and recover to check if some data/metadata provided by matrix user corresponds to the eth address it claims to be.

**Comment by andrevmatos** [comment]
@palango what's the status of your web3.py implementation? Anything pull-requestable? I'm working on demoing a matrix-based `RaidenProtocol` implementation, but would like to integrate it with web3 already. Any change of you sharing your work on this?

**Comment by palango** [comment]
All progress is here: https://github.com/palango/raiden/commits/web3-ification2

But I didn't spend more time on it.

**Comment by carver** [comment]
We just released web3 [beta 13](http://web3py.readthedocs.io/en/latest/releases.html#v4-0-0-beta-13), I highly recommend upgrading.

I took a brief look at these changes:
https://github.com/raiden-network/raiden/compare/master...palango:web3-ification2#diff-f21cc0dc64f5cbf240a380453b9be7c2R72

One thing that jumps out at me is the contract invocations, like:

```py
self.web3_proxy.functions.registerEndpoint.transact(endpoint)
```

(Where `self.web3_proxy` is a contract at `discovery_address`)

---

The way the new function invocations in web3.py work is like this:

```py
contract.functions.contractMethod(contract_args_here).transact(transaction_metadata)
```

So to transfer 3 token atoms to someone at 2 a Gwei gas price, the line would be:
```py
my_token.functions.transfer("0x1111111111111111111111111111111111111111", 3).transact({'gasPrice': 2 * 10**9})
```

Even if your contract function doesn't take any arguments, it must be called, like so:
```py
self.web3_proxy.functions.registerEndpoint().transact(endpoint)
```
_(I didn't see the ABI, but I suspect that `endpoint` is supposed to be an argument to `registerEndpoint`)_

---

Please let me know if you all get stuck on something with web3py. I'm happy to help.

**Comment by palango** [comment]
Hej @carver,

thanks for the heads-up. Especially the replacement for `deploy` is nice, as that gave us lots of warnings in tests.

This branch was a check to see how much work porting would incur, so I didn't know how to use the contract functions back then. I'll try to find some time to update this.

**Comment by palango** [comment]
As we sign all transactions locally further progress here is blocked by ethereum/web3.py#649

**Comment by carver** [comment]
Technically, everything that's possible on https://github.com/ethereum/web3.py/pull/649 is already possible without it. It's just a convenience thing. You can sign with `w3.eth.account.signTransaction()` and then use `w3.eth.sendRawTransaction()`: http://web3py.readthedocs.io/en/stable/web3.eth.account.html#sign-a-transaction

So it doesn't _have_ to be a blocker, but if you want to wait for the more concise API :+1:

**Comment by palango** [comment]
Yes, this is more an architectural problem and I didn't describe that really well.

We have wrapper classes in raiden that use our home-build contract proxies. When replacing these proxies I run into the problem that the `transact` call (like in `self.web3_proxy.functions.registerEndpoint().transact(endpoint)`) requires an unlocked account in the ethereum node. So it would be easiest to wait for locally signed transactions so we do not have to change the current architecture here.

**Comment by palango** [comment]
Closing this in favor of the follow up issues.
